src/session.c

```diff
@@ -526,7 +550,11 @@ static CK_RV fetch_session(P11PROV_SESSION_POOL *pool, CK_FLAGS flags,
         ret = MUTEX_LOCK(pool->login_session);
         if (ret == CKR_OK) {
             if (pool->login_session->in_use) {
-                ret = CKR_CANT_LOCK;
+                if (is_login_state(pool->login_session->state)) {
```

Comment:
This has a race condition: if the other thread is still executing login, but has not completed it, this will immediately return `CKR_CANT_LOCK`. The time size of the window depends on how slow/fast the token is to complete a login, and how fast the parallel login attempt happens. This also likely happens only if two threads initiate a key search operation requiring login at exact same time.

Reply:
Yeah, that is why I still can return a CANT_LOCK error ...

Reply:
A small nit regarding using `CKR_CANT_LOCK`: it is against the specification's purpose for this error:
>  #define CKR_CANT_LOCK
> This value can only be returned by '''C_Initialize'''. It means that the type of locking requested by the application for thread-safety is not available in this library, and so the application cannot make use of this library in the specified fashion. 

A better option would perhaps be:
> #define CKR_OPERATION_ACTIVE
> There is already an active operation (or combination of active operations) which prevents Cryptoki from activating the specified operation.

Or just specify our own from the `CKR_VENDOR_DEFINED` range.